### DISCUSSION TITLE
dispatch while render?

### DISCUSSION THREAD

**Discussion by arcanis** [reduxjs/react-redux#1943] [Q&A] [answered]
The React documentation suggests that, in some cases, it's better to [call `setState` from within `render` than to do it from within an effect](https://beta.reactjs.org/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes). Is it true as well from `dispatch`?

**Comment by markerikson** [comment, marked as answer]
No, _never_.

The use case of calling `setState()` while rendering _only_ works because React special-cases that exact behavior.  If it detects that you tried to do a `setState()` while rendering, it queues up that state update and _immediately, synchronously_ re-runs the rendering for that component before continuing.  That means that conceptually this can _sorta_ be seen as "not actually a side effect", because the behavior is never visible to the outside world (similar to creating a new object/array in a function, mutating it, and then returning it - no one else ever _saw_ it being mutated, so it's okay in that case).

With Redux, on the hand, dispatching an action while rendering is most _definitely_ a side effect.  The Redux store gets updated immediately, and that is an externally visible (across the app) change.  

Moreover, dispatching actions normally causes Redux-connected components to queue a re-render... and React _will_ throw errors if component B has a render queued up while it's in the middle of rendering component A.

So, no, don't _ever_ dispatch actions in the middle of rendering a React component :)

**Reply by mrdulin** [reply]
Good explanation, should be in the documentation
